contract/src/wallet/WalletV5R1Contract.kt

```diff
@@ -0,0 +1,253 @@
+            privateKey: PrivateKeyEd25519,
+            validUntil: Int,
+            seqno: Int,
+            walletId: WalletId,
+            vararg transfers: WalletTransfer
+        ): Message<Cell> {
+            val info = ExtInMsgInfo(
+                src = AddrNone,
+                dest = address,
+                importFee = Coins()
+            )
+
+            val maybeStateInit =
+                Maybe.of(stateInit?.let {
+                    Either.of<StateInit, CellRef<StateInit>>(
+                        null,
+                        CellRef(value = it, StateInit)
+                    )
+                })
+
+
+            val transferBody = createTransferMessageBody(
+                privateKey,
+                validUntil,
+                seqno,
+                walletId,
+                *transfers
+            )
+
+            val body = Either.of<Cell, CellRef<Cell>>(null, CellRef(value = transferBody, AnyTlbConstructor))
+            return Message(
+                info = info,
+                init = maybeStateInit,
+                body = body
+            )
+        }
+
+        private fun createTransferMessageBody(
+            privateKey: PrivateKeyEd25519,
+            validUntil: Int,
+            seqno: Int,
+            walletId: WalletId,
+            vararg gifts: WalletTransfer
+        ): Cell {
+            val packed = packV5Actions(*gifts)
+            val unsignedBody = CellBuilder.createCell {
+                storeUInt(0x7369676E, 32) // MessageType.ext
+                storeUInt(walletId.serialize(), 32)
+                if(seqno == 0) {
+                    storeUInt(0xFFFFFFFF, 32)
+                } else {
+                    storeUInt(validUntil, 32)
+                }
+                storeUInt(seqno, 32)
+                storeBitString(packed.bits)
+                storeRefs(packed.refs)
+            }
+
+            val signature = BitString(privateKey.sign(unsignedBody.hash().toByteArray()))
+
+            return CellBuilder.createCell {
+                storeBitString(unsignedBody.bits)
+                storeRefs(unsignedBody.refs)
+                storeBitString(signature)
+            }
+        }
+
+        private fun packV5Actions(vararg gifts: WalletTransfer): CellBuilder {
+
+            var latestCell = Cell.empty()
+            for (gift in gifts) {
+                val intMsg = CellRef(gift.toMessageRelaxed(), MessageRelaxed.tlbCodec(AnyTlbConstructor))
+
+                latestCell = CellBuilder.createCell {
+                    storeUInt(0x0ec3c86d, 32) // OUT_ACTION_SEND_MSG_TAG
+                    storeUInt(gift.sendMode, 8)
+                    storeRefs(latestCell)
+                    storeRefs(intMsg.cell)
+                }
+            }
+
+            return CellBuilder.beginCell().apply {
+                storeBoolean(true)
+                storeRef(latestCell)
+                storeBoolean(false)
+            }
+        }
+    }
+}
```

Comment:
end of file without space

Reply:
https://github.com/ton-blockchain/ton-kotlin/pull/131/commits/f6617722c078812bb2202292d3e9eb96b56b018d